core/lz4/src/ZipLZ4.cxx

```diff
@@ -0,0 +1,139 @@
+extern "C" void LZ4_XXH64_canonicalFromHash(XXH64_canonical_t *dst, XXH64_hash_t hash);
+extern "C" XXH64_hash_t LZ4_XXH64_hashFromCanonical(const XXH64_canonical_t *src);
+
+// Header consists of:
+// - 2 byte identifier "L4"
+// - 1 byte LZ4 version string.
+// - 3 bytes of uncompressed size
+// - 3 bytes of compressed size
+// - 8 byte checksum using xxhash 64.
+static const int kChecksumOffset = 2 + 1 + 3 + 3;
+static const int kChecksumSize = sizeof(XXH64_canonical_t);
+static const int kHeaderSize = kChecksumOffset + kChecksumSize;
+
+void R__zipLZ4(int cxlevel, int *srcsize, char *src, int *tgtsize, char *tgt, int *irep)
+{
+   int LZ4_version = LZ4_versionNumber();
+   uint64_t out_size; /* compressed size */
+   uint64_t in_size = (unsigned)(*srcsize);
+
+   *irep = 0;
+
+   if (R__unlikely(*tgtsize <= 0)) {
+      return;
+   }
+
+   // Refuse to compress more than 16MB at a time -- we are only allowed 3 bytes for size info.
+   if (R__unlikely(*srcsize > 0xffffff || *srcsize < 0)) {
+      return;
+   }
+
+   int returnStatus;
+   if (cxlevel > 9) {
+      cxlevel = 9;
+   }
+   if (cxlevel >= 4) {
+      returnStatus = LZ4_compress_HC(src, &tgt[kHeaderSize], *srcsize, *tgtsize - kHeaderSize, cxlevel);
+   } else {
+      returnStatus = LZ4_compress_default(src, &tgt[kHeaderSize], *srcsize, *tgtsize - kHeaderSize);
+   }
+
+   if (R__unlikely(returnStatus == 0)) { /* LZ4 compression failed */
+      return;
+   }
+   XXH64_hash_t checksumResult = LZ4_XXH64(tgt + kHeaderSize, returnStatus, 0);
+
+   tgt[0] = 'L';
+   tgt[1] = '4';
+   tgt[2] = (LZ4_version / (100 * 100));
+
+   out_size = returnStatus + kChecksumSize; /* compressed size, including the checksum. */
+
+   // NOTE: these next 6 bytes are required from the ROOT compressed buffer format;
+   // upper layers will assume they are laid out in a specific manner.
+   tgt[3] = (char)(out_size & 0xff);
+   tgt[4] = (char)((out_size >> 8) & 0xff);
+   tgt[5] = (char)((out_size >> 16) & 0xff);
+
+   tgt[6] = (char)(in_size & 0xff); /* decompressed size */
+   tgt[7] = (char)((in_size >> 8) & 0xff);
+   tgt[8] = (char)((in_size >> 16) & 0xff);
+
+   // Write out checksum.
+   LZ4_XXH64_canonicalFromHash(reinterpret_cast<XXH64_canonical_t *>(tgt + kChecksumOffset), checksumResult);
+
+   *irep = (int)returnStatus + kHeaderSize;
+}
+
+void R__unzipLZ4(int *srcsize, unsigned char *src, int *tgtsize, unsigned char *tgt, int *irep)
+{
+   // NOTE: We don't check that srcsize / tgtsize is reasonable or within the ROOT-imposed limits.
+   // This is assumed to be handled by the upper layers.
+
+   int LZ4_version = LZ4_versionNumber() / (100 * 100);
+   *irep = 0;
+   if (R__unlikely(src[0] != 'L' || src[1] != '4')) {
+      fprintf(stderr, "R__unzipLZ4: algorithm run against buffer with incorrect header (got %d%d; expected %d%d).\n",
+              src[0], src[1], 'L', '4');
+      return;
+   }
+   if (R__unlikely(src[2] != LZ4_version)) {
+      fprintf(stderr,
```

Comment:
That's unfortunate (but likely unavoidable :( ) that we can not use the ROOT error printing here ... this means that the frameworks can not intercept those messages.